Governor fractional voting [UNI review]

davidlaprade opened the pull request:
This is the simple/naive approach to vote storage. We store the different vote types (`for`, `against`, `abstain`) as three `uint128`s packed into 2 slots. From a high level, here are some stats on the approach:
* `castVoteWithReasonAndParams` costs 92501 gas on average
* max vote count per type is ~3.4e38 which should be... enough

Here's the full gas estimation report from hardhat:

```
·----------------------------------------------------------|---------------------------|-------------|-----------------------------·
|                   Solc version: 0.8.9                    ·  Optimizer enabled: true  ·  Runs: 200  ·  Block limit: 10000000 gas  │
···························································|···························|·············|······························
|  Methods                                                                                                                         │
···························|·······························|·············|·············|·············|···············|··············
|  Contract                ·  Method                       ·  Min        ·  Max        ·  Avg        ·  # calls      ·  usd (avg)  │
···························|·······························|·············|·············|·············|···············|··············
|  ERC20VotesMock          ·  delegate                     ·          -  ·          -  ·      48386  ·           84  ·          -  │
···························|·······························|·············|·············|·············|···············|··············
|  ERC20VotesMock          ·  mint                         ·          -  ·          -  ·     118350  ·           21  ·          -  │
···························|·······························|·············|·············|·············|···············|··············
|  ERC20VotesMock          ·  transfer                     ·     103007  ·     103091  ·     103027  ·           28  ·          -  │
···························|·······························|·············|·············|·············|···············|··············
|  GovernorFractionalMock  ·  castVote                     ·      69625  ·      86833  ·      83346  ·           10  ·          -  │
···························|·······························|·············|·············|·············|···············|··············
|  GovernorFractionalMock  ·  castVoteWithReason           ·          -  ·          -  ·      88142  ·            1  ·          -  │
···························|·······························|·············|·············|·············|···············|··············
|  GovernorFractionalMock  ·  castVoteWithReasonAndParams  ·      75405  ·     112357  ·      92501  ·            9  ·          -  │
···························|·······························|·············|·············|·············|···············|··············
|  GovernorFractionalMock  ·  execute                      ·      63557  ·      63569  ·      63565  ·            3  ·          -  │
···························|·······························|·············|·············|·············|···············|··············
|  GovernorFractionalMock  ·  propose                      ·      87761  ·      87773  ·      87772  ·           17  ·          -  │
···························|·······························|·············|·············|·············|···············|··············
|  Deployments                                             ·                                         ·  % of limit   ·             │
···························································|·············|·············|·············|···············|··············
|  CallReceiverMock                                        ·          -  ·          -  ·     342850  ·        3.4 %  ·          -  │
···························································|·············|·············|·············|···············|··············
|  ERC20VotesMock                                          ·          -  ·          -  ·    1752759  ·       17.5 %  ·          -  │
···························································|·············|·············|·············|···············|··············
|  GovernorFractionalMock                                  ·    2682184  ·    2682196  ·    2682195  ·       26.8 %  ·          -  │
·----------------------------------------------------------|-------------|-------------|-------------|---------------|-------------·
```


#### PR Checklist

<!-- Before merging the pull request all of the following must be complete. -->
<!-- Feel free to submit a PR or Draft PR even if some items are pending. -->
<!-- Some of the items may not apply. -->

- [ ] Tests
- [ ] Documentation
- [ ] Changelog entry

apbendi committed:
Initial implementation of fractional voting w/ naive storage layout

* Implement fractional voting with a un-optimized internal storage layout
* Set up test harness for fractional voting, ensure nominal case is still
  functional, exercise basic fractional voting capability in tests

davidlaprade committed:
Gas golfing: reduce vote storage size

davidlaprade committed:
Gas golfing: move hasVoted to its own storage slot to reduce gas

davidlaprade committed:
Correct Address function description

davidlaprade committed:
Gas golfing: switch to uint80 to fit all votes in a single slot

davidlaprade committed:
Test fractional voting

davidlaprade committed:
yarn prettier

davidlaprade committed:
Store votes as uint128's

davidlaprade committed:
Remove unused imports and variable

davidlaprade committed:
Begin reimplementing tests in the new architecture

davidlaprade committed:
Re-implement more tests, delete old test file

davidlaprade committed:
Reimplement vote overflow testing

davidlaprade committed:
Rename test file and get rid of old one

davidlaprade committed:
yarn prettier + npm run lint contracts/governance

davidlaprade committed:
Unpack bits so that params can be submitted as a single 32 byte word

davidlaprade committed:
Pack votes in tests

davidlaprade committed:
prettier + lint

davidlaprade committed:
Incorporate PR feedback

davidlaprade committed:
Minor gas optimization tweaks

davidlaprade committed:
Incorporate PR feedback

davidlaprade committed:
Fix test broken by revert message change

davidlaprade committed:
Fix variable names

NoahZinsmeister commented on contracts/governance/extensions/GovernorCountingFractional.sol:
```diff
@@ -0,0 +1,156 @@
+        }
+    }
+
+    uint256 constant _VOTEMASK = 0xffffffffffffffffffffffffffffffff; // 128 bits of 0's, 128 bits of 1's
```
probably better to explicitly specify visibility

davidlaprade replied:
We can just get rid of this now. Thanks

NoahZinsmeister commented on contracts/governance/extensions/GovernorCountingFractional.sol:
```diff
@@ -0,0 +1,156 @@
+
+        uint256 decoded = abi.decode(voteData, (uint256));
+        uint128 forVotes = uint128(decoded >> 128); // keep left-most 128 bits
+        uint128 againstVotes = uint128(_VOTEMASK & decoded); // keep right-most 128 bits
```
shouldn't `uint128(decoded)` truncate implicitly? i guess you could argue this is more clear but i'd prefer a simple cast + comment

davidlaprade replied:
Nice, good catch!

NoahZinsmeister commented on contracts/governance/extensions/GovernorCountingFractional.sol:
```diff
@@ -0,0 +1,156 @@
+        uint128 forVotes = uint128(decoded >> 128); // keep left-most 128 bits
+        uint128 againstVotes = uint128(_VOTEMASK & decoded); // keep right-most 128 bits
+
+        require(forVotes + againstVotes <= SafeCast.toUint128(weight), "GovernorCountingFractional: invalid weight");
```
this is really nit-picky, but you could technically do `uint256(forVotes) + againstVotes <= SafeCast.toUint128(weight)` inside of an unchecked block - the upcast would ensure that the addition couldn't overflow, and the condition would still be safely enforced. doubt it saves much gas, though, and it's a bit ugly, so very much optional

davidlaprade replied:
Ah, yup, you're right! Thanks

NoahZinsmeister commented on contracts/governance/extensions/GovernorCountingFractional.sol:
```diff
@@ -0,0 +1,156 @@
+            abstainVotes = uint128(weight) - forVotes - againstVotes;
+        }
+
+        ProposalVote memory existingProposalVote = _proposalVotes[proposalId];
```
slight preference for this to be declared as `storage` and then written to, to stay as close to the `_countVoteNominal` implementation as possible

davidlaprade replied:
Yeah, good point. We had done that initially but found that it cost ~500 more gas. I guess the optimizer isn't smart enough to batch the writes. For now I'm just going to leave it as is, since it's slightly more efficient. But we'll see what OZ thinks

moodysalem replied:
i might use the same variable here, style nit though

apbendi replied:
avoided second variable declaration in d11325f3378b284cb5d013758a2e8ca7142e488f

NoahZinsmeister reviewed (commented):
looks good overall! left a few comments, but nothing hugely important

davidlaprade committed:
Modify based on PR feedback

davidlaprade commented:
Rebased on top of the OZ master as of fcf35e57 and incorporated PR feedback with [3e6e92b](https://github.com/ScopeLift/openzeppelin-contracts/pull/6/commits/3e6e92b090b275de926a4a593289e2623453a59e)

moodysalem commented on contracts/governance/extensions/GovernorCountingFractional.sol:
```diff
@@ -0,0 +1,157 @@
+        uint128 abstainVotes;
+        unchecked {
+            require(
+                uint256(forVotes) + againstVotes <= SafeCast.toUint128(weight),
```
is the safecast necessary?

davidlaprade replied:
I think so. Here's the scenario that we're worried about:

* `weight > type(uint128).max` and will overflow a regular cast to `uint128` into a large number
* forVotes + againstVotes is smaller than the incorrectly cast weight
* we end up with no revert and an `abstain` value which is incorrect

If we safecast weight, we guarantee this cannot happen.

moodysalem commented on contracts/mocks/GovernorFractionalMock.sol:
```diff
@@ -0,0 +1,31 @@
+    constructor(string memory name_, IVotes token_) Governor(name_) GovernorVotes(token_) {}
+
+    function quorum(uint256) public pure override returns (uint256) {
+        return 10e18; // 10 ether
```
remove comment imo

moodysalem reviewed (commented):
solidity looks good, just nits

davidlaprade commented on contracts/governance/extensions/GovernorCountingFractional.sol:
```diff
@@ -0,0 +1,157 @@
+    /**
+     * @dev Supported vote types. Matches Governor Bravo ordering.
+     */
+    enum VoteType {
+        Against,
+        For,
+        Abstain
+    }
```
Yeah, good question. The only other places that I see this defined are

* [GovernorCompatibilityBravo](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/14ca3aeb798d9b9be31df86ae7ef8b8f760caa4c/contracts/governance/compatibility/GovernorCompatibilityBravo.sol#L26-L30)
* [GovernorCountingSimple](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/14ca3aeb798d9b9be31df86ae7ef8b8f760caa4c/contracts/governance/extensions/GovernorCountingSimple.sol#L17-L21)

I guess it just feels a little funny to reference the enum values from those other contracts, given that we don't inherit from either of them, and that a concrete implementation that used `GovernorCountingFractional` could conflict with either/both.

But I agree that it's not ideal that this enum is duplicated in a bunch of places. Open to suggestions here!

apbendi replied:
Done in d11325f3378b284cb5d013758a2e8ca7142e488f

davidlaprade commented on contracts/governance/extensions/GovernorCountingFractional.sol:
```diff
@@ -0,0 +1,157 @@
+     */
+    // solhint-disable-next-line func-name-mixedcase
+    function COUNTING_MODE() public pure virtual override returns (string memory) {
+        return "support=bravo&quorum=for";
```
Yup, this is unfinished. That's why we left the TODO above. We were just planning to wait and get OZ's feedback on it.

Though I just realized that `quorum=for` is incorrect -- it should be `quorum=bravo` per [this](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/14ca3aeb798d9b9be31df86ae7ef8b8f760caa4c/contracts/governance/IGovernor.sol#L93). And we might as well just add `params` to reflect what is in `IGovernor` too: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/14ca3aeb798d9b9be31df86ae7ef8b8f760caa4c/contracts/governance/IGovernor.sol#L99. I'll make those updates

davidlaprade replied:
[2cf016c](https://github.com/ScopeLift/openzeppelin-contracts/pull/6/commits/2cf016c55bd327bc1d57a0e12e6619a6afafc6f6)

davidlaprade commented on contracts/governance/extensions/GovernorCountingFractional.sol:
```diff
@@ -0,0 +1,157 @@
+
+    /**
+     * @dev See {IGovernor-COUNTING_MODE}.
+     * TODO: Add param for how params is used?
```
https://github.com/ScopeLift/openzeppelin-contracts/pull/6#discussion_r864166483

davidlaprade committed:
Fix COUNTING_MODE according to IGovernor rules

davidlaprade committed:
Update based on PR feedback

davidlaprade committed:
Fix test broken by changing COUNTING_MODE

davidlaprade committed:
Pull abstain votes out of params, do not infer them from non-votes

davidlaprade committed:
Test that we can differentiate between abstaining and not voting

davidlaprade committed:
Refactor tests to get rid of some ugly BN math

apbendi committed:
Decode votes with internal method using assembly

apbendi committed:
Remove outdated TODO for counting mode params

apbendi committed:
Stylistic improvements to counting fractional
